### PR TITLE
Prepare release version

### PR DESCRIPTION
**Pull request by philipheinisch**
Includes small fixes to support all Windows power profiling libraries in a single build.
Attached is a precompiled version including the required HDF5 DLLs for testing before release.

[toolkitiCL.zip](https://github.com/IANW-Projects/toolkitICL/files/2975714/toolkitiCL.zip)

### CODE REVIEW

**Comment by ranocha** [comment]
I will test the binaries this afternoon/evening.

File: src/main.cpp [hunk 1]
```diff
@@ -944,9 +939,43 @@ int main(int argc, char *argv[]) {
 #if defined(USENVML)
   if (nvidia_log_power || nvidia_log_temp)
   {
-    cout << "Using NVML interface..." << endl;
-    h5_create_dir(out_name, "/housekeeping");
-    h5_create_dir(out_name, "/housekeeping/nvidia");
+      //Get Program Files path from system
+      TCHAR pf[MAX_PATH];
```
Comment (ranocha) [inline src/main.cpp:941]: This needs additional include guards and/or definitions on Linux:
```
toolkitICL/src/main.cpp:943:7: error: ‘TCHAR’ was not declared in this scope
       TCHAR pf[MAX_PATH];
       ^~~~~
/home/hendrik/Research/AIKEF/toolkitICL/src/main.cpp:944:32: error: ‘pf’ was not declared in this scope
       SHGetSpecialFolderPath(0,pf,CSIDL_PROGRAM_FILES,FALSE);
                                ^~
/home/hendrik/Research/AIKEF/toolkitICL/src/main.cpp:944:35: error: ‘CSIDL_PROGRAM_FILES’ was not declared in this scope
       SHGetSpecialFolderPath(0,pf,CSIDL_PROGRAM_FILES,FALSE);
                                   ^~~~~~~~~~~~~~~~~~~
[ 67%] Linking CXX executable copy_char
/home/hendrik/Research/AIKEF/toolkitICL/src/main.cpp:944:35: note: suggested alternative: ‘CL_PROGRAM_IL’
       SHGetSpecialFolderPath(0,pf,CSIDL_PROGRAM_FILES,FALSE);
                                   ^~~~~~~~~~~~~~~~~~~
                                   CL_PROGRAM_IL
/home/hendrik/Research/AIKEF/toolkitICL/src/main.cpp:944:55: error: ‘FALSE’ was not declared in this scope
       SHGetSpecialFolderPath(0,pf,CSIDL_PROGRAM_FILES,FALSE);
                                                       ^~~~~
/home/hendrik/Research/AIKEF/toolkitICL/src/main.cpp:944:55: note: suggested alternative: ‘FILE’
       SHGetSpecialFolderPath(0,pf,CSIDL_PROGRAM_FILES,FALSE);
                                                       ^~~~~
                                                       FILE
/home/hendrik/Research/AIKEF/toolkitICL/src/main.cpp:944:7: error: ‘SHGetSpecialFolderPath’ was not declared in this scope
       SHGetSpecialFolderPath(0,pf,CSIDL_PROGRAM_FILES,FALSE);
       ^~~~~~~~~~~~~~~~~~~~~~
/home/hendrik/Research/AIKEF/toolkitICL/src/main.cpp:949:11: error: ‘LoadLibraryEx’ was not declared in this scope
           LoadLibraryEx(nvsmi_path.c_str(), NULL, 0);
           ^~~~~~~~~~~~~
/home/hendrik/Research/AIKEF/toolkitICL/src/main.cpp:952:15: error: ‘LoadLibraryEx’ was not declared in this scope
               LoadLibraryEx("nvml.dll", NULL, 0);

```

**Review by ranocha — CHANGES_REQUESTED** [review]

**Comment by ranocha** [comment]
I get an error on Windows using the binaries if I select power/temperature logging for Intel:
```
toolkitICL> .\toolkitICL.exe -intel_power 500 -intel_temp 500 -d 2 -b -c SSPRK_test.h5
Available OpenCL devices: 3
Benchmark mode

GeForce GTX 1070 Ti
OpenCL version: OpenCL 1.2 CUDA
Memory limit: 8589934592
WG limit: 1024

Reading kernel from file: SSPRK_kernel.cl ...
Found Kernels: 4
Number of Kernels to execute: 46608
Ingesting HDF5 config file...
Creating output HDF5 file...
Old HDF5 data file found and deleted!
Setting range...
Using Intel Power Gadget interface...
error: failed to initialize MDH_Context
Sleeping for 4s

Launching kernel...
Kernels executed: 46608
Kernel runtime: 2342 ms

Sleeping for 4s
Saving results...
```
Nevertheless, the logging seems to be fine.
```julia
julia> h5open("out_SSPRK_test.h5", "r") do io
           display(read(io, "/housekeeping/intel"))
       end
Dict{String,Any} with 8 entries:
  "temperature_time"    => [1.55291e9, 1.55291e9, 1.55291e9, 1.55291e9, 1.55291e9, 1.55291e9, 1.55291e9, 1.55291e9, 1.5…
  "TDP"                 => 95.0
  "package_temperature" => UInt16[0x0022, 0x0022, 0x0022, 0x0022, 0x0022, 0x0022, 0x0022, 0x0032, 0x0036, 0x0036  …  0x…
  "power_time"          => [1.55291e9, 1.55291e9, 1.55291e9, 1.55291e9, 1.55291e9, 1.55291e9, 1.55291e9, 1.55291e9, 1.5…
  "DRAM0"               => Float32[0.444122, 0.459119, 0.45039, 0.446602, 0.448259, 0.446364, 0.449206, 0.463428, 0.508…
  "Package_power_limit" => 95.0
  "cores0"              => Float32[1.58965, 1.83813, 1.23916, 1.19193, 1.22519, 1.16007, 1.5695, 7.0671, 29.8957, 30.12…
  "package0"            => Float32[4.97291, 5.25945, 4.66966, 4.61355, 4.65522, 4.57991, 4.99975, 10.4579, 33.3535, 33.…
```

**Comment by ranocha** [comment]
Apart from the error message `error: failed to initialize MDH_Context`, everything seems to be fine on both Windows and Linux.

File: src/main.cpp (hunk 1 above)
Comment (philipheinisch) [reply src/main.cpp:941]: Somehow an `#if defined(_WIN32)` got lost. This should be fixed now.

**Review by ranocha — APPROVED** [review]

**Comment by philipheinisch** [comment]
I get the same thing on all systems, which are not the original build system. It seems to me, that somehow the interface gets reinitialized - which fails, but as the library is already loaded, it still works. Do we want to try and fix this prior to release?

**Comment by ranocha** [comment]
Since the error does not seem to be crucial and all functionality is available, we might publish this version now. However, if you think that it is not too difficult or time consuming for you to fix this issue now, that would of course be nice.

File: src/main.cpp [hunk 2]
```diff
@@ -333,14 +333,11 @@ void intel_log_temp_func()
       if (rapl->detect_socket1() == true)
       {
         temp1 = rapl->get_temp1();
+		intel_temp1.push_back(temp1);
```
Comment (ranocha) [inline src/main.cpp:336]: Visual Studio?

File: src/rapl.cpp [hunk 3]
```diff
@@ -131,7 +131,7 @@ Rapl::Rapl(){
     pGetNumNodes(&numCPUnodes);
     if (numCPUnodes > 1)
     {
-        pp1_supported = true;
+		socket1_detected = true;
```
Comment (ranocha) [inline src/rapl.cpp:134]: Visual Studio?

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -59,10 +59,13 @@ add_executable(toolkitICL ${SOURCES} $<TARGET_OBJECTS:hdf5_io>)
 
 set(LIBRARIES "${OpenCL_LIBRARIES};${HDF5_HL_LIBRARIES};${HDF5_LIBRARIES}")
 
+#set(MSVC_LINK_FLAGS "/DELAYLOAD:")
+
 # link libraries
 IF(USENVML)
   if(MSVC)
     list(APPEND LIBRARIES "nvml.lib")
+	set(MSVC_LINK_FLAGS "${MSVC_LINK_FLAGS}/DELAYLOAD:nvml.dll ")
```
Comment (ranocha) [inline src/CMakeLists.txt]: Visual Studio?

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -71,11 +74,20 @@ ENDIF(USENVML)
 IF(USEAMDP)
   if(MSVC)
     list(APPEND LIBRARIES "AMDPowerProfileAPI.lib")
+	set(MSVC_LINK_FLAGS "${MSVC_LINK_FLAGS}/DELAYLOAD:AMDPowerProfileAPI.dll ")
```
Comment (ranocha) [inline src/CMakeLists.txt]: Visual Studio?

**Pull request merged**